### ISSUE TITLE
[BUG] Does not play video ? (Format: hls(scalable or drm or GnuTLS) )

### ISSUE THREAD

**Issue by lavara123** [jhomlala/betterplayer#309]
**Describe the bug**
better_player doesn't play this but plays mpv player

**To Reproduce**
```dart
    BetterPlayerDataSource betterPlayerDataSource = BetterPlayerDataSource(
      BetterPlayerDataSourceType.network,
      "https://dbx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3/sheila",
      videoFormat: BetterPlayerVideoFormat.hls,
      headers: {
        "Referer":
            "https://ydx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3",
        "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0"
      },
    );
```

**Expected behavior**
should play because it plays mpv player (Gives an error but still plays)

https://user-images.githubusercontent.com/12005386/108413417-e9173980-7222-11eb-9c75-1b68a408b372.mp4

it is mpv player: https://github.com/mpv-player/mpv

**mpv.sh**
```bash
mpv --ytdl-raw-options=config-location="dene.conf" "https://dbx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3/sheila"
```

**dene.conf**
```bash
--add-header "Referer: https://ydx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3"
--add-header "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0"
```
![Ekran görüntüsü_2021-02-18_22-59-51](https://user-images.githubusercontent.com/12005386/108414267-ef59e580-7223-11eb-8b61-01c6062c6d02.png)

and run:
```bash
bash mpv.sh
```
**Flutter doctor**
[✓] Flutter (Channel stable, 1.22.6, on Linux, locale tr_TR.utf8)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)

**Better Player version**
- Version: ^0.0.55

**Additional context**
![Ekran görüntüsü_2021-02-18_23-07-59](https://user-images.githubusercontent.com/12005386/108415158-12d16000-7225-11eb-8f62-a4bb4f606248.png)
link to video link: 
[click..](https://yabancidizi.pw/dizi/avatar-the-last-airbender-izle/sezon-1/bolum-1)

**Comment by jhomlala** [comment]
@lavara123 Let's be clear - Better Player (and other mobile players too) won't play all possible videos that you want. If something can't be played by Android ExoPlayer or iOS AvPlayer then I can't help you. 

Anyway when I tried to run your video I saw this error: 
```
PlatformException(VideoError, Failed to load video: You do not have permission to access the requested resource., null, null)
```
So seems like Better Player can't access your video since it protected.

**Comment by jhomlala** [comment]
Also when I visit https://dbx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3/sheila I can't access your video.

**Comment by lavara123** [comment]
> Also when I visit https://dbx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3/sheila I can't access your video.

This is inaccessible referer and User-Agent required. 
Try the mpv example and you will see that it is accessible

**Comment by jhomlala** [comment]
I don't have motivation to download external players and use them. From my perspective, there's something wrong with your server configuration. 

Check this example:

```
 void checkSource() async {
    try {
      String url =
          "https://dbx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3/sheila";
      var headers = {
        "Referer":
            "https://ydx.molystream.net/embed/27756-602d4e94a12d7f654d26a6f3",
        "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0"
      };
      HttpClient _httpClient = HttpClient()
        ..connectionTimeout = const Duration(seconds: 5);
      final request = await _httpClient.getUrl(Uri.parse(url));
      if (headers != null) {
        headers.forEach((name, value) => request.headers.add(name, value));
      }

      final response = await request.close();
      var data = "";
      await response.transform(const Utf8Decoder()).listen((contents) {
        data += contents.toString();
      }).asFuture<String>();

      print("Data: " + data);
    } catch (exception) {
      print("Failed: $exception");
      return null;
    }
  }
```
It simply connects to url of your data source and try to get content of this url. 
Result:
```
error code: 1020
```

Again, double check everything with your configuration. On Better Player side everything is fine.

**Comment by lavara123** [comment]
> It simply connects to url of your data source and try to get content of this url.
> Result:
> 
> ```
> error code: 1020
> ```
> 
> Again, double check everything with your configuration. On Better Player side everything is fine.

After waiting 5 seconds, the content returns 😲😲
![Ekran görüntüsü_2021-02-19_02-51-41](https://user-images.githubusercontent.com/12005386/108437121-964e7980-7244-11eb-812a-37e5ca9dcf60.png)


better_player can't play it this way ?

**Comment by lavara123** [comment]
@jhomlala It is enough to wait 500 milliseconds for this url, so can the connectionTimeout feature be added to the player?

**Comment by jhomlala** [comment]
@lavara123 Again - the problem is on your side. Your resource can't be accessed by video player and you must setup your video on different server.

I've tried again checkSource() method and result was same:
```dart
error code: 1020
```
I don't know how did you get correct data of HLS. Maybe this server check IP address.

Anyway I've checked custom loading policy in ExoPlayer with same result - 403.

```java
 return new HlsMediaSource.Factory(mediaDataSourceFactory)
                        .setLoadErrorHandlingPolicy(new LoadErrorHandlingPolicy() {

                            @Override
                            public long getRetryDelayMsFor(LoadErrorInfo loadErrorInfo) {
                                return 1000;
                            }

                            @Override
                            public int getMinimumLoadableRetryCount(int dataType) {
                                return 5;
                            }
                        })
                        .createMediaSource(MediaItem.fromUri(uri));
```

Right now I have no clue why it's not working. For me 403 means that you're not providing correct data and request can't be authorised. Need for wait/delay is an indicator that's something is wrong here.

Feel free to debug this issue on your side, but please remember that if you want to create PR, you need to provide same handling on iOS.

**Comment by lavara123** [comment]
@jhomlala  Does this work flawlessly on you? https://reqbin.com/c-ne9atvvc

![Ekran görüntüsü_2021-02-19_23-26-43](https://user-images.githubusercontent.com/12005386/108557966-3366ec00-72f1-11eb-88db-53d4fe2ed199.png)

```
URL: https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09/moly
Referer: https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09
User-Agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Mobile Safari/537.36
```

better_player won't play this either

**Comment by jhomlala** [comment]
Yes, I can access this resource. 
Anyway there's still issue with your server. I've opened network profiler in Android Studio to check requests which were made by ExoPlayer.

1st request to access master playlist was fine (https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09/moly)
2nd request to access child playlist was fine(https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09/q/1)
3rd request to access elements of child playlist failed. Check screenshots below.
ExoPlayer is sending requests correctly with provided headers.

Again, please change your server configuration. It doesn't work correctly. 

<img width="602" alt="Screenshot 2021-02-20 at 14 32 46" src="https://user-images.githubusercontent.com/1329033/108597291-95befb80-7388-11eb-8b44-81f28f6708a8.png">
<img width="612" alt="Screenshot 2021-02-20 at 14 32 39" src="https://user-images.githubusercontent.com/1329033/108597290-948dce80-7388-11eb-9ede-7901c47c27c9.png">
<img width="606" alt="Screenshot 2021-02-20 at 14 32 33" src="https://user-images.githubusercontent.com/1329033/108597288-9192de00-7388-11eb-956f-3445f0756035.png">

**Comment by lavara123** [comment]
@jhomlala Does not connect different country :/ The code you used to play this URL?

In my country, the return is successful
But better_player doesn't play anyway

**Comment by lavara123** [comment]
@jhomlala 

my calls work fine

```dart
import 'dart:io';

void checkSource() async {
  try {
    var url =
        'https://yd1.kavga5.cf/v/27601-602fb891abd5b61889921d09/480p/moly_000.png';
    var headers = {
      'Referer':
          'https://yd.sheila.stream/embed/28026-5e8c2b513adcea42a6644e7c',
      'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0'
    };
    var _httpClient = HttpClient();
    final request = await _httpClient.getUrl(Uri.parse(url));
    if (headers != null) {
      headers.forEach((name, value) => request.headers.add(name, value));
    }

    final response = await request.close();
    await response.pipe(File('out_0.png').openWrite());

    print('Data: ' + File('out_0.png').openRead().toString());
  } catch (exception) {
    print('Failed: $exception');
    return null;
  }
}
```
![Ekran görüntüsü_2021-02-20_20-40-18](https://user-images.githubusercontent.com/12005386/108603959-e04d7180-73a2-11eb-9f04-4fe67b64adcc.png)

I guess the problem is related to better or exo

**Comment by lavara123** [comment]
> <img alt="Screenshot 2021-02-20 at 14 32 33" width="606" src="https://user-images.githubusercontent.com/1329033/108597288-9192de00-7388-11eb-956f-3445f0756035.png">

hey, 3rd child won't play with http headers

I think this is the fault

**Comment by lavara123** [comment]
It is also important that the user-agent is correct
![Ekran görüntüsü_2021-02-20_22-16-12](https://user-images.githubusercontent.com/12005386/108606127-2fe66a00-73b0-11eb-8473-79848220daf9.png)
I'm working with this User-Agent=ExoPlayer
```bash
ffmpeg -referer "https://yd.sheila.stream/" -user_agent "ExoPlayer" -i "https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09/moly" -f hls -vcodec libx264 -crf 23 -preset veryfast  -c:a copy - | ffplay -
```
same error = 403
![Ekran görüntüsü_2021-02-20_22-14-39](https://user-images.githubusercontent.com/12005386/108606159-55737380-73b0-11eb-82a2-f308801272a3.png)

no problem if i run it with a proper User-Agent = `Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0`
works without error
```bash
ffmpeg -referer "https://yd.sheila.stream/" -user_agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0" -i "https://yd.sheila.stream/embed/27601-602fb891abd5b61889921d09/moly" -f hls -vcodec libx264 -crf 23 -preset veryfast  -c:a copy - | ffplay -
```
![Ekran görüntüsü_2021-02-20_22-26-26](https://user-images.githubusercontent.com/12005386/108606348-928c3580-73b1-11eb-9b50-dd7746529793.png)

**Comment by jhomlala** [comment]
> > <img alt="Screenshot 2021-02-20 at 14 32 33" width="606" src="https://user-images.githubusercontent.com/1329033/108597288-9192de00-7388-11eb-956f-3445f0756035.png">
> 
> hey, 3rd child won't play with http headers
> 
> I think this is the fault

Once you've pass headers it will be used by ExoPlayer for all master/child playlist items. Maybe you can try fork Better Player and create custom data source with ResolvingDataSource, where you'll opt-in headers in http requests: https://exoplayer.dev/customization.html

**Comment by lavara123** [comment]
> > hey, 3rd child won't play with http headers
> > I think this is the fault
> 
> Once you've pass headers it will be used by ExoPlayer for all master/child playlist items (and this is normal behaviour of ExoPlayer). Maybe you can try fork Better Player and create custom data source with ResolvingDataSource, where you'll opt-in headers in http requests: https://exoplayer.dev/customization.html

                            https://github.com/jhomlala/betterplayer/blob/8ad1afe109060ba053c275e2e6f095952bf782f2/android/src/main/java/com/jhomlala/better_player/BetterPlayer.java#L131

**Comment by lavara123** [comment]
@jhomlala The problem is definitely user-agent, can you please fix this?

https://github.com/jhomlala/betterplayer/blob/8ad1afe109060ba053c275e2e6f095952bf782f2/android/src/main/java/com/jhomlala/better_player/BetterPlayer.java#L131

https://github.com/jhomlala/betterplayer/blob/8ad1afe109060ba053c275e2e6f095952bf782f2/android/src/main/java/com/jhomlala/better_player/BetterPlayer.java#L148

I make "ExoPlayer" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0"

works well that way

![Screenshot_20210221-134835185](https://user-images.githubusercontent.com/12005386/108622701-6fee3100-7432-11eb-9873-12b7e5fe3d03.jpg)

**Comment by jhomlala** [comment]
Fixed user agent issue in 0.0.57. Video can't be played on my side. Hope that this fix will help you.

**Comment by lavara123** [comment]
> Fixed user agent issue in 0.0.57. Video can't be played on my side. Hope that this fix will help you.

Thank you very much, it works for me 👍
